libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h

```diff
@@ -0,0 +1,213 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA__MEMORY_RESOURCE_CUDA_MEMORY_RESOURCE_H
+#define _CUDA__MEMORY_RESOURCE_CUDA_MEMORY_RESOURCE_H
+
+#include <cuda/__cccl_config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#if !defined(_CCCL_CUDA_COMPILER_NVCC) && !defined(_CCCL_CUDA_COMPILER_NVHPC)
+#  include <cuda_runtime_api.h>
+#endif // !_CCCL_CUDA_COMPILER_NVCC && !_CCCL_CUDA_COMPILER_NVHPC
+
+#include <cuda/__memory_resource/get_property.h>
+#include <cuda/__memory_resource/properties.h>
+#include <cuda/__memory_resource/resource_ref.h>
+#include <cuda/__memory_resource/resource.h>
+
+#if _CCCL_STD_VER >= 2014
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA_MR
+
+/**
+ * @brief `cuda_memory_resource` uses cudaMalloc / cudaFree for allocation/deallocation.
+ */
+struct cuda_memory_resource
+{
+  /**
+   * @brief Allocate device memory of size at least \p __bytes.
+   * @param __bytes The size in bytes of the allocation.
+   * @param __alignment The requested alignment of the allocation. Is ignored!
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* allocate(const size_t __bytes, const size_t __alignment) const
+  {
+    _LIBCUDACXX_ASSERT(__alignment <= 256 && (256 % __alignment == 0),
```

Comment:
suggestion: I think allocation should throw, not assert. Also note that this assertion/error means that the alignment parameter is NOT ignored.

Reply:
Question: Just because cudaMalloc aligns to 256 (or possibly 512, even though documented at 256), doesn't mean an MR can't support larger allocations. I suppose users who *really* want that can write a wrapping resource that aligns, but is there another reason we don't support it here?  Seems a bit strange that cuda::mr has this parameter and our most basic implementation semi-ignores it and adds another method without it that is not part of the concept.

Reply:
> Seems a bit strange that cuda::mr has this parameter and our most basic implementation semi-ignores it and adds another method without it that is not part of the concept.

The existence of the alignment parameter does not mean all possible alignment values are valid. Each resource implementation will have its own limitations and this includes alignment. 

`cuda_memory_resource` is intended to be the simplest possible wrapper on top of `cudaMalloc`, and so it will inherit all the limitations of `cudaMalloc`, including alignment. 

We _could_ add an additional resource that allows for larger alignments, (or an adapter!), but that would no longer be a trivial wrapper on top of `cudaMalloc`.